### PR TITLE
AgendaItems(Task and Reminder) detailScreen

### PR DESCRIPTION
**Pull request by Gerondk**

### CODE REVIEW

File: agenda/data/src/main/java/com/gkp/agenda/data/OfflineAgendaRepository.kt [hunk 1]
```diff
@@ -37,6 +37,18 @@ class OfflineAgendaRepository(
...
+            taskyRetrofitApi.getTask(id).toAgendaTaskItem()
+        }
+    }
+
+    override fun fetchAgendaItemReminder(id: String): Flow<TaskyResult<AgendaItem.Reminder>> {
+        return networkApiCall {
+            taskyRetrofitApi.getReminder(id).toAgendaReminderItem()
+        }
+    }
+
```
Comment (Kai-Cilliers) [inline agenda/data/src/main/java/com/gkp/agenda/data/OfflineAgendaRepository.kt:51]: `Flow`s make more sense when you expect a continuous flow of values that need to be observed (like observing changes in user location updates or changes in a database table). In these cases a normal `suspend` function makes more sense when returning a single result.

File: agenda/presentation/src/main/java/com/gkp/agenda/presentation/detail/AgendaItemDetailScreen.kt [hunk 2]
```diff
@@ -27,24 +27,37 @@ import com.gkp.agenda.presentation.compoments.AgendaItemDescription
...
+) {
+    val viewModel = koinViewModel<AgendaItemDetailViewModel>()
+    val state = viewModel.uiState
+
+    AgendaItemDetailScreen(
+        onClickBackButton = onClickBackButton,
+        onEditClick = {
+            navigateToTaskEdit(state.id, state.agendaItemType)
+        },
+        onClickDeleteButton = {},
```
Comment (Kai-Cilliers) [inline agenda/presentation/src/main/java/com/gkp/agenda/presentation/detail/AgendaItemDetailScreen.kt:50]: Did you mean to reference your `onClickBackButton` lambda here?

File: agenda/presentation/src/main/java/com/gkp/agenda/presentation/home/AgendaUiState.kt [hunk 3]
```diff
@@ -16,4 +20,26 @@ data class AgendaUiState(
 
     val uiDates: List<DateWithSelected>
         get() = getCurrentAndNextSixWeekDayDates(selectedDate)
+
+    val uiFormattedSelectedDate: String
+        @Composable
+        get() = when (selectedDate) {
+            LocalDate.now() -> {
+                stringResource(R.string.today)
+            }
```
Comment (Kai-Cilliers) [inline agenda/presentation/src/main/java/com/gkp/agenda/presentation/home/AgendaUiState.kt]: I like to use the following super helpful class to manage text and string resources. With it, I can use it within my data classes and ViewModel, and when inside my composables or with `Context`, I can then fetch the actual resource.

```
sealed interface UiText {
    data class DynamicString(val value: String): UiText
    class StringResource(
        @StringRes val id: Int,
        val args: Array<Any> = arrayOf()
    ): UiText

    @Composable
    fun asString(): String {
        return when(this) {
            is DynamicString -> value
            is StringResource -> stringResource(id = id, *args)
        }
    }

    fun asString(context: Context): String {
        return when(this) {
            is DynamicString -> value
            is StringResource -> context.getString(id, *args)
        }
    }
}
```

File: agenda/presentation/src/main/java/com/gkp/agenda/presentation/home/AgendaUiState.kt [hunk 4]
```diff
@@ -16,4 +20,26 @@ data class AgendaUiState(
 
     val uiDates: List<DateWithSelected>
         get() = getCurrentAndNextSixWeekDayDates(selectedDate)
+
+    val uiFormattedSelectedDate: String
+        @Composable
+        get() = when (selectedDate) {
```
Comment (Kai-Cilliers) [inline agenda/presentation/src/main/java/com/gkp/agenda/presentation/home/AgendaUiState.kt]: Composable functions are tied to the Compose runtime and its lifecycle and I am not sure how this would behave when used inside a data class like this 🤔 This annotation also ties this data class directly to Compose, which makes it very coupled and makes testing harder.

See my comment on the `UIText` interface and see how you can use it to modify this class not to be tied to Compose directly.

**Review by Kai-Cilliers — APPROVED** [review]

File: agenda/presentation/src/main/java/com/gkp/agenda/presentation/detail/AgendaItemDetailViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    var uiState by mutableStateOf(AgendaItemDetailUiState())
+        private set
+
+    init {
+        val detailScreenRoute = savedStateHandle.toRoute<AgendaItemDetailScreenRoute>()
+        val itemId = detailScreenRoute.taskId
+        val itemType = detailScreenRoute.agendaItemType
+
+        if (itemId.isNotBlank()) {
```
Comment (philipplackner) [inline agenda/presentation/src/main/java/com/gkp/agenda/presentation/detail/AgendaItemDetailViewModel.kt]: Passing in blank IDs to indicate they don't exist is wrong design. Non-nullable types communicate that they are mandatory and always exist. If you then bypass this by passing in empty invalid IDs, this is everything but obvious. Nullable types are used for exactly that: Indicating a value may or may not exist. If you encounter a nullable field somewhere in a codebase, you know there might be cases it isn't there (which you don't know if a field is non-nullable but can still not exist). 

Long story short: I'd recommend to make the ID nullable

File: agenda/presentation/src/main/java/com/gkp/agenda/presentation/home/AgendaUiState.kt (hunk 4 above)
Comment (philipplackner) [reply agenda/presentation/src/main/java/com/gkp/agenda/presentation/home/AgendaUiState.kt]: I'm not 100% sure, but coupling this to Compose might prevent you from writing JVM unit tests for your ViewModel (that uses this state class). JVM unit tests may have no references to the Android SDK - now I'm not sure if Compose falls under that, but either way it does couple the state to Compose, that's true. I'd also prefer the UiText approach from below

**Review by philipplackner — APPROVED** [review]
